js/lib/AlphabeticIndex.js

```diff
@@ -221,128 +353,303 @@ AlphabeticIndex.prototype.addLabels = function(labels, start) {
+ *         label: "B",
+ *         elements: ["B", "Baaa", "Boo"]
+ *     },
+ *     ...
+ *     {
+ *         label: "#",
+ *         elements: ["3par.com", "@handle"]
+ *     }
+ * ]
  * </code>
- * 
+ *
  * All elements within a bucket are sorted per the collation
  * for the locale of this index.
- * 
+ *
  * @returns {Object} a hash of all buckets and elements
  * as per the description above.
  */
 AlphabeticIndex.prototype.getAllBuckets = function() {
+	var underflowIndex = -1, overflowIndex = -1;
+	var mixedScriptEndIndex = -1;
+	var count = 0;
+	var temp;
+	var i;
+
+	this._updateCollationMap(this.collation.default);
+
+	this.index.sort(function(a,b){
+		if (a.label < b.label) {
+			return -1;
+		} else {
+			return 1;
+		}
+	});
+
+	for (i=0; i < this.index.length; i++) {
+		if (this.inherit &&
+			this.indexUnits.indexOf(this.index[i].label) == -1) {
+			mixedScriptEndIndex = i;
+			count++;
+		}
+	}
+
+	if (this.inherit && count > 0) {
+		temp = this.index.splice((mixedScriptEndIndex - count) +1 , count);
+		this.index = this.index.concat(temp);
+	}
+
+	for (i=0; i < this.index.length; i++) {
+		if (this.index[i].label == this.underflowLabel) {
+			underflowIndex = i
+			break;
+		}
+	}
+
+	if (underflowIndex > 0) {
+		temp = this.index.splice(underflowIndex,1)[0];
+		this.index.unshift(temp);
+	}
+
+	for (i=0; i < this.index.length; i++) {
+		if (this.index[i].label == this.overflowLabel) {
+			overflowIndex = i
+			break;
+		}
+	}
+
+	if (overflowIndex > -1) {
+		temp = this.index.splice(overflowIndex,1)[0];
+		this.index[this.index.length] = temp;
+	}
+
+	return this.index;
 };
 
 /**
- * Return the label of the bucket for a given element. This 
+ * Return the label of the bucket for a given element. This
  * follows the rules set up when the index was instantiated to
  * find the bucket into which the element would go if it
  * were added to this index. The element is not added to
  * the index, however. (See addElement for that.)
- * 
+ *
  * @param {String} element the element to check
- * @returns {String} the label for the bucket for this element
+ * @returns {String|undefined} the label for the bucket for this element
  */
 AlphabeticIndex.prototype.getBucket = function(element) {
+	var label;
+	var firstChar;
+	var collationValue;
+	var charNum, firstBoundaryChar, endBoundaryChar, firstCharNum, endCharNum;
+	
+	if (!element && !typeof(element) === 'string') {
+		return undefined;
+	}
+
+	firstChar = element.charAt(0);
+
+	if (CType.withinRange(firstChar, "hangul")) {
+		firstChar = this._normalizeHangul(firstChar);
+	}
+
+	collationValue = this.collationMap[firstChar];
+
+	if (!collationValue && this.inherit) {
+		for (var i = 0; i < this.inherit.length; i++) {
+			if (this.inherit[i] === 'this') {
+				this.inherit[i] = this.style;
+    		}
+
+    		if (this.collation[this.inherit[i]].map[firstChar]) {
+    			collationValue = this.collation[this.inherit[i]].map[firstChar];
+    			this.inheritStyle = this.inherit[i];
```

Comment:
This is changing an instance variable with a temporary value. For safety, temporary values should be passed as parameters to other functions/methods so that they do not change if some other code calls a different method. Because JS is single-threaded, it is hard to get it wrong, but when you store temporary values in an instance variable and multiple methods can change it, then the value of the variable depends on which method was called last.

Much better would be to create a super-set map during initialization so that this work only has to be done once. For example, let's say a collation for hangul inherits latin. (inherit: ["this", "latin"]) Hangul has primary values that go from 27 to 238 (211 values). Latin has primary values that go from 0 to 69 (70 values). We build a new map with primary values that go from 0 to 280 and store that as an instance variable that doesn't change for the life of the instance. (Values 0 to 210 are Hangul and 211 to 280 are Latin) Then, when we want to look up a value, we check the one single map...

Reply:
I got your considerations. 
In order to create one single map for dealing mixed script case, There's one more issue which is 'flowBoundaries' Every collation map has different values...  and What if they have the same value on both maps? such as double Backslash(\\). it is in 'ascii' and latin'